cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java

```diff
@@ -573,8 +576,32 @@ public void destroy() {
       }
     }
 
-    LineageOperationsProcessor processor = new LineageOperationsProcessor(spec.getConnections(), allStageOperations,
-                                                                          noMergeRequiredStages);
+    if (allStageOperations.isEmpty()) {
+      return;
+    }
+
+    Map<String, StageInputOutput> stageInputOutputs = new HashMap<>();
+    for (Map.Entry<String, StageSpec> stageSpecEntry : stageSpecs.entrySet()) {
+      Set<String> inputFields = new HashSet<>();
+      Set<String> outputFields = new HashSet<>();
+      Schema outputSchema = stageSpecEntry.getValue().getOutputSchema();
+      if (outputSchema != null && outputSchema.getFields() != null) {
+        outputFields.addAll(outputSchema.getFields().stream().map(Schema.Field::getName).collect(Collectors.toSet()));
+      }
+
+      Map<String, Schema> inputSchemas = stageSpecEntry.getValue().getInputSchemas();
+      for (Map.Entry<String, Schema> entry : inputSchemas.entrySet()) {
```

Comment:
Assume that you have customer schema as 
```java
|id | first_name | last_name |  street_address      |   city    | state | zipcode | phone number 
```
and purchase schema as

```java
| customer_id | item   | price |
```
and Join plugin properties are

```java
"properties": {
        "selectedFields": "customers.id as customer_id,customers.first_name as name,purchases.item,purchases.price",
        "requiredInputs": "customers, purchases",
        "joinKeys": "customers.id = purchases.customer_id"
    }
```
Then the field lineage operations would be 
```java
join: (id, first_name, last_name, street_address, city, state, zipcode, phone number, customer_id, item, price) -> (id, first_name, item, price)
rename_id: id -> customer_id
rename_first_name: first_name -> name 
```
Even if the fields have same name coming in from two different stages, it can be represented as above. However there is JIRA for having specialized JOIN operation https://issues.cask.co/browse/CDAP-13298 where it might be possible to represent JOIN more granular level when fields from different inputs have same name.